### PR TITLE
keep track of linearizer times in process_replay

### PR DESCRIPTION
**Pull request by DKormann**
re: https://github.com/tinygrad/tinygrad/pull/5093#issuecomment-2183598699

will show if linearizer delay changes like this:
```
Perf:   3.27 ->   3.64 ns r_256_256                                                                                                                                            | 0/1 
Perf:   1.34 ->   3.18 ns r_2560_2560
100%|█████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 106.05it/s]
```

### CODE REVIEW

**Comment by DKormann** [comment]
will only print if difference is more than 10% and more than 10 ms

**Comment by DKormann** [comment]
time results are super off for some kernels must be missing some caching or optimization

**Review by Qazalin — COMMENTED** [review]
I like the idea of doing it in process replay. I think you can add an acceptable/noise perf diff threshold that the PR shouldn't regress above.

**Comment by DKormann** [comment]
> I like the idea of doing it in process replay. I think you can add an acceptable/noise perf diff threshold that the PR shouldn't regress above.

thats what im doing with the 10% rule. or what do you mean?

but right now i'm still measuring wrong things. It show me 50 -> 0 and vice versa repeatadly so i mus be missing something

**Comment by Qazalin** [comment]
I mean asserting it as something that'd fail process replay if it's a regression. speedup is fine though.

**Comment by DKormann** [comment]
ok integrated in given assert / fail structure
set acceptable slowdown to 20%

**Comment by Qazalin** [comment]
I need to see this work in action.
Can you try adding a time.sleep in your branch and compare with master and show it works?

I tried testing it against a regression and it didn't work like I expected
https://github.com/tinygrad/tinygrad/compare/master...linearizer_speed_replay

process replay run with real regression: https://github.com/tinygrad/tinygrad/actions/runs/9633535581/job/26568067013#step:17:30
without, it still fails: https://github.com/tinygrad/tinygrad/actions/runs/9633702896

**Review by Qazalin — COMMENTED** [review]
I need to see this work in action.
Can you try adding a time.sleep in your branch and compare with master and show it works?

I tried testing it against a regression and it didn't work like I expected
https://github.com/tinygrad/tinygrad/compare/master...linearizer_speed_replay

process replay run with real regression: https://github.com/tinygrad/tinygrad/actions/runs/9633535581/job/26568067013#step:17:30
without, it still fails: https://github.com/tinygrad/tinygrad/actions/runs/9633702896

File: tinygrad/codegen/linearizer.py [hunk 1]
```diff
@@ -523,9 +523,14 @@ def to_program(self) -> Program:
...
     src = self.opts.render(to_function_name(self.name), self.uops)
-    if getenv("RUN_PROCESS_REPLAY"): diskcache_put("process_replay", id(self), (self, src))
+    if getenv("RUN_PROCESS_REPLAY"): diskcache_put("process_replay", id(self), (self, src, time_lin(self)))
     ops, mem = self.uops.flops_mem()
     run_count = prod((self.global_size if self.global_size else []) + (self.local_size if self.local_size else []))
     # NOTE: we use min here to ignore the indexing FLOPS
     return Program(self.name, src, self.opts.device, self.global_size, self.local_size,
                    self.uops, min(info.flops, ops * run_count), min(info.mem_estimate, mem * run_count))
+
+def time_lin(lin:Linearizer):
```
Comment (Qazalin) [inline tinygrad/codegen/linearizer.py:533]: might be worth adding a `timeit` in tinygrad.helpers

**Comment by DKormann** [comment]
should be better now
tested time.sleep(0.1) in uopsgraph.linearize

```
➜  tinygrad git:(linearizer_speed_replay) ✗ python test/external/process_replay/process_replay.py 
  0%|                                                                                                                                                  | 0/9 [00:00<?, ?it/s]TIME REPLAY DETECTED CHANGE 0.0008254999993368983 ->0.10634350002510473
(LazyOp(op=BufferOps.STORE, src=(LazyOp(op=BinaryOps.ADD, src=(LazyOp(op=ReduceOps.SUM, src=(LazyOp(op=BufferOps.CONST, src=(), arg=ConstBuffer(val=1, dtype=dtypes.int, st=ShapeTracker(views=(View(shape=(2561, 5119), strides=(0, 0), offset=0, mask=((0, 2561), (2559, 5119)), contiguous=False), View(shape=(2560, 2560), strides=(1, 5120), offset=0, mask=None, contiguous=False))))),), arg=(1,)), LazyOp(op=BufferOps.CONST, src=(), arg=ConstBuffer(val=-1, dtype=dtypes.int, st=ShapeTracker(views=(View(shape=(2560, 1), strides=(0, 0), offset=0, mask=None, contiguous=False),))))), arg=None),), arg=MemBuffer(idx=0, dtype=dtypes.int, st=ShapeTracker(views=(View(shape=(2560, 1), strides=(1, 0), offset=0, mask=None, contiguous=True),)))),)
[]
  0%|                                                                                                                                                  | 0/9 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "/Users/wittgenstein/code/dkormann/tinygrad/test/external/process_replay/process_replay.py", line 29, in <module>
    if getenv("ASSERT_PROCESS_REPLAY", 1): raise e
  File "/Users/wittgenstein/code/dkormann/tinygrad/test/external/process_replay/process_replay.py", line 21, in <module>
    assert (time_-time_baseline)/max(time_baseline, .1) < 0.1, f"TIME REPLAY DETECTED CHANGE {time_baseline} ->{time_}"
AssertionError: TIME REPLAY DETECTED CHANGE 0.0008254999993368983 ->0.10634350002510473
```

File: tinygrad/codegen/linearizer.py [hunk 2]
```diff
@@ -523,9 +523,11 @@ def to_program(self) -> Program:
     self.linearize()
     info = get_lazyop_info(self.ast[0])
     src = self.opts.render(to_function_name(self.name), self.uops)
-    if getenv("RUN_PROCESS_REPLAY"): diskcache_put("process_replay", id(self), (self, src))
+    if getenv("RUN_PROCESS_REPLAY"): diskcache_put("process_replay", id(self), (self, src, time_lin(self)))
```
Comment (Qazalin) [inline tinygrad/codegen/linearizer.py:526]: can you make this a lambda

File: tinygrad/helpers.py [hunk 3]
```diff
@@ -279,4 +279,9 @@ def fn(x): return (f"{scl(x):.{3-math.ceil(math.log10(scl(x)))}f}"[:4]+(f"{[' ',
     print(bar[:term+1],flush=True,end='\n'*close,file=sys.stderr)
 
 class trange(tqdm):
-  def __init__(self, n:int, **kwargs): super().__init__(iterable=range(n), total=n, **kwargs)
\ No newline at end of file
+  def __init__(self, n:int, **kwargs): super().__init__(iterable=range(n), total=n, **kwargs)
+
+def timeit(fn:Callable):
```
Comment (Qazalin) [inline tinygrad/helpers.py:284]: I think the timeit API should get *args and **kwargs

File: test/external/process_replay/process_replay.py [hunk 4]
```diff
@@ -11,16 +11,19 @@
...
-    compare_k, compare_src = pickle.loads(row[0])
+    compare_k, compare_src, time_baseline = pickle.loads(row[0])
     k = Linearizer(*compare_k.ast, opts=compare_k.opts)
     for opt in compare_k.applied_opts: k.apply_opt(opt)
     good_src = k.opts.render(to_function_name(compare_k.name), k.linearize().uops)
-    try: assert compare_src == good_src
+    time_ = time_lin(k)
+    try:
+      assert compare_src == good_src, "PROCESS REPLAY DETECTED CHANGE"
+      assert (time_-time_baseline)/max(time_baseline, .1) < 0.1, f"TIME REPLAY DETECTED CHANGE {time_baseline} ->{time_}"
```
Comment (Qazalin) [inline test/external/process_replay/process_replay.py:21]: Let's remove this assert for now and just print - We need to stress test flakiness before making it a merge blocker.

File: test/external/process_replay/process_replay.py [hunk 5]
```diff
@@ -11,16 +12,18 @@
 for offset in tqdm(range(0, row_count, page_size)):
   cur.execute(f"SELECT val FROM 'process_replay_{VERSION}' LIMIT ? OFFSET ?", (page_size, offset))
   for row in cur.fetchall():
-    compare_k, compare_src = pickle.loads(row[0])
+    compare_k, compare_src, time_baseline = pickle.loads(row[0])
     k = Linearizer(*compare_k.ast, opts=compare_k.opts)
     for opt in compare_k.applied_opts: k.apply_opt(opt)
     good_src = k.opts.render(to_function_name(compare_k.name), k.linearize().uops)
-    try: assert compare_src == good_src
+    time_ = timeit(UOpGraph(k.uops).linearize)
```
Comment (Qazalin) [inline test/external/process_replay/process_replay.py:19]: hmm This is wrong actually, k.uops is the entire linearized version
Why don't you just time k.linearize().uops?

**Comment by DKormann** [comment]
i like creating a new UOpgraph since it will reset the UOpgraph.nodes which acts as a cache for dedup

**Pull request closed without merging**